### ISSUE TITLE
F401 false positive with PEP-695 upper bound specification

### ISSUE THREAD

**Issue by bryanforbes** [astral-sh/ruff#8017]
Running the following command with Ruff `0.1.0`:

```
ruff --target-version py312 --select F401 --isolated --no-cache blah.py
```

Given the following code:

```python
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable

    from .foo import Record

type RecordCallback[R: Record] = Callable[[R], None]


def process_record[R: Record](record: R) -> None:
    ...


class RecordContainer[R: Record]:
    def add_record(self, record: R) -> None:
        ...
```

`F401` is reported on line 8, but `Record` is being used in the upper bound specifications.

**Comment by charliermarsh** [comment]
Strange, will take a look.

**Comment by charliermarsh** [comment]
This is a regression in the most recent release due to #7968.
